### ISSUE TITLE
Unexplained "Aborted(OOM)" after processing 100+ images

### ISSUE THREAD

**Issue by josephrocca** [kleisauke/wasm-vips#13]
Is there some sort of memory freeing function that I need to call after processing an image? [Here's a playground link](https://wasm-vips.kleisauke.nl/playground/?deflate=hVl7j9y2Ef8qi0MB3yH2nt4PJ3aRpAlqtGiLtPkrFyASHyIliqL4kEQG_u7l3p3ts5GbHLC7t6vfvIfDGfKXq-vOeIlO1Elk-SKvb06_38lT_BPEnpwW5vTm9MvdFbNWmde3t_ysCcZnbm9xV4p2M2Wi9iI9j2q4u3r5R7g5V3wpj12FLoVwpJDFQTHSMq0hXCM8Y40QDQoVhKuO-VhSM-x7AeJIUui2Ut7VFYhb5FYFu2ycdw2Ea4c5eGvtusD6rTOvOtvJTOwlhNtRNsg28KoZQbmonl06s-TIYdxYzYbwvaIdAeXWfd9I1Qy4zcH4OllPTPSyVSUYX0eE7vW-eNGD8Q2dzPmgjWcE9F8I5cZoS8S-gHYUUzaiUbp2tCC_YzG2GxapJjgPgk_Xssqk6TAot7Kc17gxvBIgLhsI1XVoRjblEG5rx3Ih5ljrBNRvClqXXbpg60F-SMl582ZPthaMx0oZHRtBonSQH6F8mw7vt86B-vE5YUVtvLcKxM2HnDHJXF8EMP-KfelauoqZzPC6XCu_73XeVgy0V-OkU91uKsFBnFodQ9m2T4GDfllmGZys0JQikB9vSlJqLccAr8tKdi4Yu1sVQH5dMImJBXorq8hPyedwnkxItANrihn0sw57ucfwsjCD-pWkoUki_YRzEDduNOm7bedYw_wQqkidOLcnoL074-Oc-Hyf64h73t4N6YrxSuEwZqB-Q4LKpNSTTMD4VrrhZc3mMKWgHaot87EyRI4tvA-yuW0JwgUbQbnUu7EZNtLsG7iOGAs86Rdm3AHqJ6UtWeOobQjoZ5Oy6aiXnBsG5oua0thY5IQbDOonET827Le20R4yN5sOZzcxSQuKLTFpm7bbeAwzWHY3Xgx8KNtjAHEu0Xsu9mjICqdpFnx70CH0sPvytcE2PcqyuLjl-TS1m82VM_0RKLh86xHXa72NuYPToJ6RVdgKf4zg8hi8ym2WlX6Dl1GVqEG3k5Y5hbfVptzm0fNWweUlLLa0NFVHYUC_sEMXqePJ7jrQz2yWth8Ot2oK4ohoZj3XrK8EKNd1YRWkl7GogvZ2PFe1WA0OsH6TlY0LDAkKL_MiWJoPzNikBPOPzqFdG12btQbzZbDONZmkroDLOOqqdBQsrSzcbhzUKeFrLv0GxpeLLMyTq6bdgHJj_0rbLVxSFfRLbo7N5RqH5E_KVfSdO3qhtxQs97IqcNroYPIFLM_s2FsenPXCg_w6v86YsjpLS7i-TJ2o2kGWbQn6b2wNN0O2KtWC_IqlNKImLTMCtGPo1eDEjoMaEyhf4va8at1nmVtBex2nedKqSTULaEdNQhG6I1kOAuK2loxFjcm8OTC-7cRoJ8VWWQy2f00h4_AWNtQIEJdnybbbrCynGtRP1shIvRxp5kCc56jeRNORcQfzuSl801VZEosvqN9sF81NWW_JAcbDHL4p2pWHWoL54pchHT0baQO32aboK7MNhTg-jTOnP1wgeCr68rAufBL8h0A7TEvepKxvNtgzrkNzU89ZqwIEC7VJ5ZqjBu5zkpbu3Tr7wjWgY4TwZdzSnR0syG_Ol7Hq5YY4PMdPYR2CoCWWFZjQmx6kD5WtMNx3ykbpfEYr0vAcFUeUzfQbLvYJ3NBZPRTbWk-Z6-A-dpnK2Lovfm_hDSQUfEmJpC28gOPAmA5Zyhy-zClAQR0q6jPNxOHhc5UGq6KYeZKMIL_8KKzJm0plC1zY4rlVzVtaTJdG53n_JbjFKG-KfILnnnI_WtMtOO8kPM9wRnSyTy2H5_OjLvDsDCY1BnFCebLRlDXVDOJwPfOM2ZTGQQ7CpYjOq4-d0wCfWzBKi4RPa5N7qL8fybHSrj9Glcb96HmcLYyvpJmbCd4HcTzQTLVJDyLBcDjfJe2GF4cux0jA2DirqmOrROOlfwbKi89WlowpV3AZoiGwpExynsL7Jc-2hOx2by2Obn5ev7aqF78OU1EW8HEYnaspq_xGLOjnzjAWDyTSdm_BNHBt71yempRTuL-Podi3QI8KjptsVieSIBJUgzjVbenGDG8FA_XbJEnndakp6kH9qF7N3Ptmx7B-qqPGH1OVeHhbEJV2a2in0sd5BZo_mjXZkOJrmoJy-0U3hfVjWzTwMRcS49zFPRDDx6KzsNVGAqsy-DiMh0Io2qPFeNCOKuQpC11CHIfLX4EG0sWW0sDH8l0TKue6TMoELM9TEhqE0SIHDdpRL_U0oK1duILt4GVRz9M4xukCjIdsqrwViDn4uE5ag9NBFCIN8HEYigW1TI9hKEF7sUo6PQWa0RW040CETpnJsYT73TgjY8-XdjUF6D-aaTEse1lkCK6TNc2StFHsgI8n521AQqFWsY9t069ff7p5kv_RCyLGEBzvn5LHB3TR14_XUqeF3t9Ofbyu-kDYi6WPJN3ecXuixCJ2_dsH6ZbMatGd9q80MVZzZAl-hRZtXqnYyfvzIHjEn2fy18j6zV9-j-_vf3t5-l0TSrQm-vXd1d3V-5uzZURe69Obtyd9vsi7vrl51PDy95kGPUcuvn4ihgfyrcTfE2mJ_l4v6vqCe0oXc9g86P_zT_-MDOL7GWnSWfLvfiTIxu_P0PB5iHi8IDdH9o9EPwhy-Xb9Ij598ZQmfj8bjSLFo6wvnlluBYlPo_VPnnzkPhD7yPo7_w5fv1icVc6-uDl3SpFoIOMCX0c-X6q5CHIWy3D9KbJfffXyIuQD8P3Dx1Ps3dWP7_717r9__-Fvd1cX2Pub68vHnfz80vJPvfz0TpPP0baNK3N-N3cDOUuy_6iX-TtHY5CvH6MWic6d1p1__Pk-wA88bm9PD3JOZjmZuRMi5tLJcExO3JyyrFDH60_SYqJF6I8dimNblJs-2sppdNE5nn1bdnobVTozwgdmP8vmL0gj53h38gj84LITEYb8OdG9oK9PH_38qMTFE_HpA831U9KY9BPRkojX9576x_3_53snv__MFSj6-GSXi6Ajvp5YLgi1kf8nO19dQDe32ZNVbiPxA-TBrC8xHzS8SIkrn9qXF5KXF9D9280HTS7MYiI-RCvSxKCefubSNt9egnivg-aW_G95jOeLSwV78XHZamKdlvdU313W8y8fef0aPXGyXpHXp7iQYr7cXghP906Ijrx6efUN5tuJ4zd3Vw8r4e7q7Te38be390_j26__Bw) that produces an OOM error after processing (resize+crop) 131 images on my machine using Chrome. [Here's a version](https://wasm-vips.kleisauke.nl/playground/?deflate=hVl7j9y2Ef8qi0MB7yH2Wu-HE7tI0gQ1WrRF2vyVCxCJpERKJEXxIYkM_N3LvYftGPZkgdvbXf3mPRzOkL_cnDvjJToNTiLLFnm-Pf1-J0_xxYk9Oc3N6fXpl7sbaq0yr16-ZBdNML4w-xJ3JW83UyZqL9LLpMa7m-efw4lcsaU8dhW6FMKRQhbHgJGWaQ3hGu4pbThvUKggXHWIY0nNuO8FiCNJodtKeVdXIG6RWxXssjHWNRCuHUXw1tp1gfVbBas628mM7yWE21E2yjawqplAuagWLhU0OXIYN1XCELZXQ0dAuXXfN1I1I25zML5O1jPlvWxVCcbXEa57vS-e92B8QydzNmrjKQH9F0K50aElfF9AO4o5m9AkXTtZkN-xGNuNi1QznAfBp2tZZdJ0GJRbWcZq3BhWcRCXjWTQdWgmOucQbmunciHmWOsE1G8OWpddumDrQX5ISbF5sydbC8ZjHegwNZxE6SA_MrBtPrzfOgfqx0RCi9p4bxWIE4cUmGSuLwKYf8W-dO2wckEEvC7Xyu97nbcVBe3VOOlUt5uKMxCnVkdRtu1zYKBfFiGDkxWaUwTyY01JSq3lFOB1WcnOBWN3qwLIrwsmMbFAb2UV-Sn5JZwnM-LtSJtCgH7WYS_3GF4aBKhfSZohSaSfcQ7ipm1I-m7bGdYwP4QqUifO7Qlo707ZJBKf76KOuC_buyFdUVYpHKYM1G9MUJmUepYJGN9KN6ysqQhzCtqh2jKfKkPk1ML7IBVtSxAu6ATKHbybmnEjzb6B64jSwJJ-ocYdoH5S2pI2brANAf1sUjof9ZIzQ8F8UXMaG4ucMINB_SRix4b91jbaQ-Zm8-HsxmdpQbElJm3TdhuLYQbL7saKkY1le4wgziV6z_keDVnhNM2Cb49hDD3svnxtsE2PsiyubvlymtrN5sqZ_ggDuHzrCddrvU25g9OgFsgqbLk_JnB5jF7lNstKv8HLqErUqNtZy3yAt9Wm3MTkWavg8hIWW9ohVUdhQL_QQxepY8nuOtDPVEjbj4db9QDiCG-EFjXtKw7KdV1YOellLKqgvR3LVc1XgwOs32xl4wJFfICXeRHskI_U2KQE828QoV0bXZu1BvNltM41mRxcAZdx1FXpxGlaWbjdOAanuK-Z9BsYX8azIGZXzbsB5cb-dWi3cE1V0C-5OTaXaxySPylX0Xfu6LneUrDcy6rAaaODyRewPNNjb1lw1nMP8uv8KvBA6ywt4foyd7xqR1m2Jei_qTXMjNmqVAvyK5bS8Jq01HDQjrFXo-M7DmpKoHyJ2_OqdZ9lbgXtdWzIk1bNqllAO2oSitAdyXIQELe1ZCpqTMTmwPi2Mx06ybfKYrD9awoZh7ewoYaDuDxLtt1mZTnXoH6yRkbq5UgzB-I8Q_XGm45MO5jPTeGbrsqSWHxB_YRdNDNlvSUHGA9z-KZoVxZqCeaLX8Z08nQaGrjNNkVfmW0s-PFhnDl9doHguejLw7rwQfBngXacl7xJad9ssGdch0RTi6xVAYKF2qRyzVED9zlJO-zdKnzhGtAxnPsybunOjhbkJ_Jlqnq5IQbP8XNYx8CHEssKTOhNj9KHylYY7jtlo3Qu0Io0PEfFEWUz_YaLfQY3dFqPxbbWc-Y6uI9d5jK27ovfW3gDCQVbUiKHFl7AcWBMxyylDl_nFKCgjtXgM0354eFzlQarohAsSSaQX34U1uRNpbIFLmzx3Kpm7VDM10bny_5LcItR3hT5DM895X60pltw3kl4nmGU6GSfWwbP50ddYOEMJjUGcVx5sg0pbSoB4nAtWEZtOsRBDsKlaBCrj53TCJ9b0GEoEjavTe6h_n4ixzp0_TGpNO5HX8bZwvhKGtHM8D6I44Fmqk16EAmGw_kuaTe8OHQ9RgLGRqGqjq4STdf-GSgvPltpMqVMwWVoCIEmZZKzFN4vWbYlZLd7a3F085f1a6t68es4F2UBH4cNopqzym_Egn7uDKXxQCJt9xZMA9f2zuWpSdkA9_cxFPsWhqOC4yab1fEk8ATVIE51W7pRw1pOQf02SVKxLvWAelC_Qa9G9L7ZMayf6gbjj7lKPLwt8Eq7NbRz6eO8As0fzZpsSLE1TUG5_aKbwvqpLRr4mAvxSXRxD8TwsajgttpIoFUGH4exUHA19GgxHrSjCnlKQ5cQx-DyV6CRdLGlNPCxfNeEyrkukzIBy_OchAZhtMhRg3bUSz2PaGsXpmA7WFnUYp6mOF2A8ZBNlbccUQcf10lrcDrygqcBPg5DsaCW6TGOJWgvVkmn5zBkwwracSAyzJnJsYT73TgjY8-WdjUF6L8h03xc9rLIEFwn6yFL0kbRAz6eFNuIuEKtou_bpl-__nDzJP-jF0SMITjePyWPD7q9Y_YUnwhmyKXj_Hy9oLqITl0_nF6_ebq7ur40sU7LkyT7E8XjVddZ334CfRLa86WP4h7EDMQiev7tSXNLhFp0p_0LTYzVDFmCX6BFmxcqTgH-MnIW8RdB_hpVef2X3-P7u9-en37XZCBaE_3q7ubu5t3txVIiz_qqgL5c5Z1vbx-te3r9QYueIRf_fiKGBfKtxN8TaYn-Xi_qfMV9ShvXgXmw4-ef_hmZxPcL0qSz5N_9RJCN3wE6JsZIgxfkRBTzSPgDJ9dv52fx6bNP6eJvF6NRpHqU-ZnnlllOIiJ65JOn7yWNxD6K-c6_xedni7PK2We3l04pEo2mjONz5PU5tRdOLnwZzx8y5quvnl-FfQx-9_Tl3Xt3f0x7d_Pj23-9_e_ff_jb3c31-bvb8_Xfnfzj5eifRuPju1Mmos0bU-byVnQjucRE_DHm4XduiAlxfoxuJLp0Wnf-8ed77R54vHx5epBzMsvJiJjtMe9OhmFyYuaUZYU6Xn2QFpMyQn_sUBwPo9z00Ug2RLdd4hm7pac3UaULJWyk9r2iDwvlD6SRc7yjeQQ-ee1EuCF_TnQv6OvTI9GTJfeeiE8faM4fk8YFMhMtCX9176l_3H--3Dv5PmTvXYGij092uQo64t9HlnMy2Mj_g50vrqDbl9lH1cRG4gfIg1mfYp40vEo5Xxk-v5I8v4Lu326fNLkyi8n5EK1Ic60uPzNpm2-vQbzXQTNL_rc8xvPZtVI-e59zH9Wk765r_5f3vH6NnjhZr8irU1xoMV9eXgkf8jY68ub5zTeYbSeGX9_dPKyOu5s337yMv725fxrffv0_) that downloads the images in parallel - just to make it faster to reproduce (**Edit**: This seems like it causes fetch to return 503 errors sometimes - might be hitting reddit's rate limit. Maybe best to stick with the serial version). Here's the code for the serial-download version in case the link stops working for whatever reason (just processes some random images a reddit crawl):
```js
(async function() {
    let urls = ["https://i.redd.it/da5l9vs50pw41.jpg","https://i.redd.it/m3pio5xwpza11.jpg","https://i.redd.it/e4n4xfdcrn171.jpg","https://i.redd.it/8lyhh8ll8cz61.jpg","https://i.redd.it/6xmxo1sgww461.jpg","https://i.redd.it/e04r96pyu7661.jpg","https://i.redd.it/onv6ztoviia81.jpg","https://i.redd.it/9gmzytttqoz61.jpg","https://i.redd.it/qmi6atan2lw51.jpg","https://i.redd.it/wc2gn9zi68j81.jpg","https://i.redd.it/c7mu1mh0x3j81.jpg","https://i.redd.it/j6mseiw6fae51.jpg","https://i.redd.it/7bb8np8gd9341.jpg","https://i.redd.it/un7khlbn9p511.jpg","https://i.redd.it/uelrbrwoylb71.jpg","https://i.redd.it/zan3igrsyhe61.jpg","https://i.redd.it/zz5vhf9elwo51.jpg","https://i.redd.it/4k2jcjnu9jt61.jpg","https://i.redd.it/xostagonpk661.jpg","https://i.redd.it/zy1q562nsad51.jpg","https://i.redd.it/6tii7d8si6l51.jpg","https://i.redd.it/2gefr7z8jhk31.jpg","https://i.redd.it/v9j5oesxq7061.jpg","https://i.redd.it/kzrr5a1odty31.jpg","https://i.redd.it/cpnmvysw0v971.jpg","https://i.redd.it/qfhfj8lej5o31.jpg","https://i.redd.it/efivkxyyvau61.jpg","https://i.redd.it/im0h47syytp61.jpg","https://i.redd.it/mxnmde2ub4z41.jpg","https://i.redd.it/4woa9fqlmem81.jpg","https://i.redd.it/9q6yww7396h71.jpg","https://i.redd.it/rd0apaws6li71.jpg","https://i.redd.it/pquhc2vwkzi31.jpg","https://i.redd.it/omnzun6ck1c71.jpg","https://i.redd.it/i85e5rrnjze51.jpg","https://i.redd.it/6nauzstwtpz71.jpg","https://i.redd.it/azs0s9vsv5651.png","https://i.redd.it/yekcl9gh84m41.jpg","https://i.redd.it/rzw5wyyvhzm51.jpg","https://i.redd.it/5e8f00nykd351.jpg","https://i.redd.it/jvf0bavwidr51.jpg","https://i.redd.it/5cc6e70uuw071.jpg","https://i.redd.it/whijm0y3wm771.png","https://i.redd.it/vcr6hi6pdzj21.jpg","https://i.redd.it/jg0c505rkn031.jpg","https://i.redd.it/6r8i57hmzk151.jpg","https://i.redd.it/p953j6senj911.jpg","https://i.redd.it/ehm99ecd4hj31.jpg","https://i.redd.it/fyuj8gve8wv61.jpg","https://i.redd.it/hhzi0bohsux51.jpg","https://i.redd.it/nnt5h8uft8e71.jpg","https://i.redd.it/s1hkx7o3ish41.jpg","https://i.redd.it/pk1ync3eisd61.jpg","https://i.redd.it/ncixvdyv98ry.jpg","https://i.redd.it/f2kxutvlknt41.jpg","https://i.redd.it/5de989avicr61.jpg","https://i.redd.it/2vi4gig59xg61.jpg","https://i.redd.it/u0rw3lwc3eq51.jpg","https://i.redd.it/52zy9xfgzbe71.jpg","https://i.redd.it/3q8dt1x554d61.png","https://i.redd.it/tvt3pusbxzf51.png","https://i.redd.it/7jd7q7vj3uv61.jpg","https://i.redd.it/7mctpdtlyxj71.png","https://i.redd.it/gyp3t225yv771.png","https://i.redd.it/60pgr9krn3f61.jpg","https://i.redd.it/x85vmjyi9pm41.jpg","https://i.redd.it/zot5tf1px4s61.png","https://i.redd.it/hxr41ui0wua71.jpg","https://i.redd.it/hmntbgxuqrf71.jpg","https://i.redd.it/el8mrm7hb6l61.png","https://i.redd.it/uazqlebns0s61.jpg","https://i.redd.it/ai3p7lqsdza71.jpg","https://i.redd.it/ktn8uzhclfj31.jpg","https://i.redd.it/4ztf3ghst0551.jpg","https://i.redd.it/fmz9q8r7sq751.png","https://i.redd.it/gtuu82nfu4m51.jpg","https://i.redd.it/ca61jlh16t061.jpg","https://i.redd.it/xfuply7inyv41.jpg","https://i.redd.it/il2zmku6kws51.jpg","https://i.redd.it/0x3f9vzpdtl31.jpg","https://i.redd.it/3sxvu3rdz0e71.jpg","https://i.redd.it/sghsuxblrv121.jpg","https://i.redd.it/n64d18rzs3o11.jpg","https://i.redd.it/hxw9izutyly21.jpg","https://i.redd.it/ayqmdfh721561.jpg","https://i.redd.it/ukal69gn59541.jpg","https://i.redd.it/j9sisg2qpp961.jpg","https://i.redd.it/4o5sl7e9hsl11.jpg","https://i.redd.it/gbpgulwdzpj01.png","https://i.redd.it/idrqrrb22uq21.jpg","https://i.redd.it/uif309pkp8o41.jpg","https://i.redd.it/7ez4zax0oxe41.jpg","https://i.redd.it/v9ej47demvu71.jpg","https://i.redd.it/9khfanlv6td81.jpg","https://i.redd.it/84ntovzvc8l81.jpg","https://i.redd.it/320vwt255k741.jpg","https://i.redd.it/n7csnrox12u41.jpg","https://i.redd.it/yic7vl8aejw31.jpg","https://i.redd.it/84y8a6203t281.jpg","https://i.redd.it/mtoris57v0x21.jpg","https://i.redd.it/sxy849qiz7n61.jpg","https://i.redd.it/yog1jyhjf8z41.jpg","https://i.redd.it/s4b6svg4lxl51.jpg", "https://i.redd.it/hdk4b5xtuzn61.jpg", "https://i.redd.it/tgko381hb8v31.jpg","https://i.redd.it/8uacm87m29pz.jpg","https://i.redd.it/8z7s1nq3c8x51.jpg","https://i.redd.it/09fwaqmy4u861.jpg","https://i.redd.it/lly5zy9utgt51.jpg","https://i.redd.it/m3oj6bnvciw51.jpg","https://i.redd.it/kzqgzlf5dn671.jpg","https://i.redd.it/vrgnyz6t6d031.jpg","https://i.redd.it/n8pr3mcqcrp61.jpg","https://i.redd.it/2vwvsbvd4wk71.png","https://i.redd.it/h7g4vq7k2ua51.jpg","https://i.redd.it/pok5vf0oyw951.jpg","https://i.redd.it/cz4io1enf9u71.jpg","https://i.redd.it/yva1g21hud651.jpg","https://i.redd.it/ag6fy2rhlxy61.jpg","https://i.redd.it/q8dp44mi00j51.jpg","https://i.redd.it/3x4ts386p2o21.jpg","https://i.redd.it/uza17i9f4kf61.png","https://i.redd.it/0d9dc3843k351.jpg","https://i.redd.it/5wx9saod3an71.jpg","https://i.redd.it/wiher0wk9i971.jpg","https://i.redd.it/x74dmusde7d71.jpg","https://i.redd.it/lpyevf1h86m71.jpg","https://i.redd.it/d7mi2ht1f5e81.jpg","https://i.redd.it/1cfmqym7hgk31.jpg","https://i.redd.it/hff40ikq83yy.jpg","https://i.redd.it/jexqfabxjp101.jpg","https://i.redd.it/t4sy6nsm8k961.jpg","https://i.redd.it/dunc1rs1xen51.jpg","https://i.redd.it/uya09vdouco51.png","https://i.redd.it/vmp6ahqncjj71.jpg","https://i.redd.it/0y2qh0j1ipx51.jpg","https://i.redd.it/fzzh0503i1o41.jpg","https://i.redd.it/i2v0ewtw9td31.png","https://i.redd.it/967oyqgk45451.jpg","https://i.redd.it/4fm6k26yvet01.jpg","https://i.redd.it/ashh5e519w981.jpg","https://i.redd.it/u9buu31s1if41.jpg","https://i.redd.it/zs1xwvzfx6961.jpg","https://i.redd.it/n8qul0zl0c761.jpg","https://i.redd.it/pav1vhsi9lh81.jpg","https://i.redd.it/vne1mqo7fcb41.jpg","https://i.redd.it/frqsmby8wd961.jpg","https://i.redd.it/pafsyxk60yt51.jpg","https://i.redd.it/l6ruqz9k5yr41.jpg","https://i.redd.it/e8q0vcpiq1161.jpg","https://i.redd.it/bor84tyj94881.jpg","https://i.redd.it/9cljma8x5do51.jpg","https://i.redd.it/mlt6vezh62h71.jpg","https://i.redd.it/iz4lpfbcosy41.jpg","https://i.redd.it/6z31hza0eui51.jpg","https://i.redd.it/54cgeaqrrs461.jpg","https://i.redd.it/a8z6uua2nn021.jpg","https://i.redd.it/k0z8cdcongr71.jpg","https://i.redd.it/7o7kgcv9oip41.jpg","https://i.redd.it/6i547mkjjvu31.jpg","https://i.redd.it/bn8639lchue51.jpg","https://i.redd.it/ntsd1gl4l1z51.jpg","https://i.redd.it/jc1cf51xgg521.jpg","https://i.redd.it/dp0arkzf2fq41.jpg","https://i.redd.it/xcefk2s3dnl81.jpg","https://i.redd.it/bgxdyio9qs471.jpg","https://i.redd.it/f2rlgow542c51.png","https://i.redd.it/v7f2018phxz71.jpg","https://i.redd.it/mvgclpc9phw51.jpg"];
    let nProcessed = 0;
    for(let url of urls) {
        let blob = await fetch(`https://temporary-restricted-cors-proxy.glitch.me?url=${url}`, {referrer:""}).then(r => r.blob());
        blob = await bicubicResizeAndCenterCrop(blob);
        const blobURL = URL.createObjectURL(blob);
        const img = document.createElement('img');
        img.src = blobURL;
        img.title = url;
        document.getElementById('output').appendChild(img);
        console.log(nProcessed++, url);
    }
    console.log("FINISHED");
})();

async function bicubicResizeAndCenterCrop(blob) {
    let im = vips.Image.newFromBuffer(await blob.arrayBuffer());

    // Resize so smallest side is 224px:
    let resizeFactor = 1;
    if(im.width > im.height) {
        resizeFactor = 224/im.height;
    } else {
        resizeFactor = 224/im.width; 
    }

    im = im.resize(resizeFactor, {kernel:vips.Kernel.cubic});

    // crop to 224x224:
    let left = (im.width - 224)/2;
    let top = (im.height - 224)/2;
    im = im.crop(left, top, 224, 224)

    let outBuffer = new Uint8Array(im.writeToBuffer('.png'));
    return new Blob([outBuffer], { type: 'image/png' });
}

```

Here's the error I'm seeing in Chrome DevTools:
```js
Aborted(OOM)
Pe @ vips.js?ad40afff8535a54512ad:1
Wa @ vips.js?ad40afff8535a54512ad:1
$func731 @ vips.wasm?ad40afff8535a54512ad:0x353b3
$mb @ vips.wasm?ad40afff8535a54512ad:0x9489
$func373 @ vips.wasm?ad40afff8535a54512ad:0x1930f
$func1635 @ vips.wasm?ad40afff8535a54512ad:0xb9d30
...
...
$func4883 @ vips.wasm?ad40afff8535a54512ad:0x281093
n.invokeEntryPoint @ vips.js?ad40afff8535a54512ad:1
self.onmessage @ vips.worker.js?ad40afff8535a54512ad:1
Show 47 more frames
vips.worker.js?ad40afff8535a54512ad:1 worker.js onmessage() captured an uncaught exception: RuntimeError: Aborted(OOM). Build with -s ASSERTIONS=1 for more info.
threadPrintErr @ vips.worker.js?ad40afff8535a54512ad:1
self.onmessage @ vips.worker.js?ad40afff8535a54512ad:1
vips.worker.js?ad40afff8535a54512ad:1 RuntimeError: Aborted(OOM). Build with -s ASSERTIONS=1 for more info.
    at Pe (vips.js?ad40afff8535a54512ad:1:5312)
    at Wa (vips.js?ad40afff8535a54512ad:1:79304)
    at vips.wasm?ad40afff8535a54512ad:0x353b3
    ...
    ...
    at vips.wasm?ad40afff8535a54512ad:0xa746
    at r (vips.js?ad40afff8535a54512ad:1:81710)
threadPrintErr @ vips.worker.js?ad40afff8535a54512ad:1
self.onmessage @ vips.worker.js?ad40afff8535a54512ad:1
vips.js?ad40afff8535a54512ad:1 worker sent an error! https://wasm-vips.kleisauke.nl/lib/vips.worker.js?ad40afff8535a54512ad:1: Uncaught RuntimeError: Aborted(OOM). Build with -s ASSERTIONS=1 for more info.
t.onerror @ vips.js?ad40afff8535a54512ad:1
error (async)
ce @ vips.js?ad40afff8535a54512ad:1
(anonymous) @ vips.js?ad40afff8535a54512ad:1
e @ vips.js?ad40afff8535a54512ad:1
t @ vips.js?ad40afff8535a54512ad:1
Promise.then (async)
(anonymous) @ vips.js?ad40afff8535a54512ad:1
Promise.then (async)
(anonymous) @ vips.js?ad40afff8535a54512ad:1
(anonymous) @ vips.js?ad40afff8535a54512ad:1
(anonymous) @ playground-runner.html:104
vips.js?ad40afff8535a54512ad:1 Uncaught ErrorEvent {isTrusted: true, message: 'Uncaught RuntimeError: Aborted(OOM). Build with -s ASSERTIONS=1 for more info.', filename: 'https://wasm-vips.kleisauke.nl/lib/vips.worker.js?ad40afff8535a54512ad', lineno: 1, colno: 1925, …}
t.onerror @ vips.js?ad40afff8535a54512ad:1
error (async)
ce @ vips.js?ad40afff8535a54512ad:1
(anonymous) @ vips.js?ad40afff8535a54512ad:1
e @ vips.js?ad40afff8535a54512ad:1
t @ vips.js?ad40afff8535a54512ad:1
Promise.then (async)
(anonymous) @ vips.js?ad40afff8535a54512ad:1
Promise.then (async)
(anonymous) @ vips.js?ad40afff8535a54512ad:1
(anonymous) @ vips.js?ad40afff8535a54512ad:1
(anonymous) @ playground-runner.html:104
vips.worker.js?ad40afff8535a54512ad:1 Uncaught RuntimeError: Aborted(OOM). Build with -s ASSERTIONS=1 for more info.
    at Pe (vips.js?ad40afff8535a54512ad:1:5312)
    at Wa (vips.js?ad40afff8535a54512ad:1:79304)
    at vips.wasm?ad40afff8535a54512ad:0x353b3
    ...
    ...
    at vips.wasm?ad40afff8535a54512ad:0xa746
    at r (vips.js?ad40afff8535a54512ad:1:81710)
```
![image](https://user-images.githubusercontent.com/1167575/159161748-ae166c7f-64b1-4d4b-b516-e726bbc02a53.png)

Any idea what's happening here?

**Comment by josephrocca** [comment]
The page seems to use up a lot of memory according to Chrome's task manager. It grows to 1.2GB usage at 131 images, and then crashes. I thought it might have had to do with creating all those PNG blobs and not revoking the URLs after processing, but removing the `URL.createObjectURL` stuff (so loop body is just `fetch` and then `bicubicResizeAndCenterCrop`) doesn't fix it.

![image](https://user-images.githubusercontent.com/1167575/159162940-723e664b-41be-4feb-b7b1-a85566d079d5.png)

**Comment by kleisauke** [comment]
Hi @josephrocca,

Currently it's required to call `delete()` on every image instance that is returned from wasm-vips. The Emscripten docs also provides some explanation for this:
https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#memory-management

It's very inconvenient with operation chains, unfortunately. The [`emscripten-auto-deletelater.patch`](https://github.com/kleisauke/wasm-vips/blob/84a86eb694805ab4cb66999c4de2d7e7d35ce859/build/patches/emscripten-auto-deletelater.patch) patch (which I still need to upstream) makes this less difficult by calling `deleteLater()` automatically during the construction of handles, you only need to do this within `Module.preRun`:
https://github.com/kleisauke/wasm-vips/blob/84a86eb694805ab4cb66999c4de2d7e7d35ce859/playground/src/playground-runner.html#L61-L64

After that, you can call `cleanup()` to clean up the used resources. For example, the playground does this before each run in order to clean up the previous run (if any):
https://github.com/kleisauke/wasm-vips/blob/84a86eb694805ab4cb66999c4de2d7e7d35ce859/playground/src/playground-runner.html#L26-L27

This is also used within unit tests, where resources are cleaned up after each test: 
https://github.com/kleisauke/wasm-vips/blob/84a86eb694805ab4cb66999c4de2d7e7d35ce859/test/unit/test_arithmetic.js#L29-L31

Given this information, I could resolve this locally by doing this within that playground link:
```diff
@@ -10,6 +10,7 @@
         img.title = url;
         document.getElementById('output').appendChild(img);
         console.log(nProcessed++, url);
+        cleanup();
     }
     console.log("FINISHED");
 })();
```

Note that I haven't documented this yet because I wasn't sure if Embind will be used in later versions, see:
https://github.com/kleisauke/wasm-vips/projects/1#card-44661960

**Comment by josephrocca** [comment]
@kleisauke Ahh, I see thanks for that. If I want to just "deal with" the chaining inconvenience so I can do the cleanup with the `bicubicResizeAndCenterCrop` function, how would I do that? The code below throws "BindingError: Object already scheduled for deletion" at `im1.delete()`.
```js
async function bicubicResizeAndCenterCrop(blob) {
    let im1 = vips.Image.newFromBuffer(await blob.arrayBuffer());

    // Resize so smallest side is 224px:
    let resizeFactor = 1;
    if (im1.width > im1.height) {
        resizeFactor = 224 / im1.height;
    } else {
        resizeFactor = 224 / im1.width;
    }

    let im2 = im1.resize(resizeFactor, { kernel: vips.Kernel.cubic });

    // crop to 224x224:
    let left = (im2.width - 224) / 2;
    let top = (im2.height - 224) / 2;
    let im3 = im2.crop(left, top, 224, 224)

    let outBuffer = new Uint8Array(im3.writeToBuffer('.png'));
    im1.delete();
    im2.delete();
    im3.delete();
    return new Blob([outBuffer], { type: 'image/png' });
}
```
Also, not on topic, but maybe not big enough for it's own issue: Should the `vips.shutdown()` method return a promise?

**Comment by kleisauke** [comment]
`EMBIND_AUTOMATIC_DELETELATER` and `delete()` are mutually exclusive. You'll need to set `vips.EMBIND_AUTOMATIC_DELETELATER = false;` in that playground if you want to control memory management using `delete()`.

> Should the `vips.shutdown()` method return a promise? (and should I be `await`ing the `delete`s above?)

I'm not sure if `vips.shutdown()` can be made asynchronous. Both calls needs probably be done synchronous.
https://github.com/kleisauke/wasm-vips/blob/84a86eb694805ab4cb66999c4de2d7e7d35ce859/src/vips-emscripten.cpp#L46-L48
https://github.com/emscripten-core/emscripten/blob/eea4056bb71493ae388a1cc22b2bda07cc4932ef/src/preamble.js#L399
https://github.com/emscripten-core/emscripten/blob/eea4056bb71493ae388a1cc22b2bda07cc4932ef/src/embind/embind.js#L1876

**Comment by josephrocca** [comment]
Ah gotcha, thanks! I wonder if this could all be made more intuitive with [`WeakRef`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef)s? In any case, it would definitely be worth adding a note about deletion stuff to the readme and playground examples, because it's unusual behavior in JavaScript land to have to worry about memory management, so I'm guessing most people would get confused by this. Thanks again!

**Comment by kleisauke** [comment]
Hopefully, in the future, wasm-vips can leverage Igalia's "Ref C++", which manages the lifetime of objects in a more convenient way.
https://github.com/Igalia/ref-cpp#destructors-and-finalizers

You're right that there's a lack of documentation on this subject in the meantime. I'll have a look at adding this, hence the "documentation"-label.

`vips.shutdown()` is only useful on non-web environments. It was added to shutdown Emscripten's thread pool, which would otherwise prevent the event loop from exiting.

Note that `EMBIND_AUTOMATIC_DELETELATER` is currently opt-in, so you don't need to set that to `false` outside the playground.

**Comment by josephrocca** [comment]
@kleisauke Looks like I still have a leak somewhere. [Here's a minimal reproduction](https://wasm-vips.kleisauke.nl/playground/?deflate=jVN_a9swEP0qh9moTDI5TTcYWRNIuhRKf2yUdP_UpXXtcyJmS0aSm3Yl330n2Um8MtgEdiLp3b137863AUvMi0whr2VqhZLAQniNJdBaC5mpNX8SlYExJOtEWPhBGxZ-aQDuhs8vZ2dXX--nN4tvl9PF2cn91_nFfDG_mC7m1xSWJ4VBwjcRBVp4LNTjLl-ONl2xOFhZW5lRFAkuymWtearK6PrcXKirGa_kMg768Aoac9Qa9QjiIA5gE3K7Qsk0jCegucvLQqdtzyW_a5WiMZgR46CVnSsNzN0Kf0g_x_BpQIv-9nq78rc5NBrxC7NZV_ajSGt6rv3VVGYnKC3qE60q5mRsDXJL5MD2Mno9eA-HgwGMx8QdQqqkUQXyQi3Zw7vXPXADokyWaKDanjxss25iuQl9E2L5pnn_0rWtzVdfHlI9vodnjopLXJ9qVc7qnGxmbaEUxROtk5f2uGNwFEFDBEaBKZOiQGPBiAxBGBgOP1bPoz1dY-NpklqyfwyHbTHOHlLC1yKzK5g4VXyFYrmyfzTiTTQlh6iD3ToDSNP2X3Geb29o15YhAR2kiWXdFG4Kf6KWWIwa5879hnvXaSC73qTkOljlOJ_p6VhRYG6Jg-oetnV_cKiQpA1bSQ5mKb5FNVX-HSbKIy94yB0jDXZu-y6278D-FXbrU7VtWklB1HG4EdJ-nroOE9MRX2thcaHabh-4j-_A99w3i1zJkLIgC_uesrs72u1auEZba-lZ3NfDbnfcd85I-1LhCA78nEeOpzFwE_SD40w8gcjGcUAhVW3jYHIc0dkklnRLr7vf) - it's basically the same code as before, but I'm just loading one image and resizing it over and over.

Accoding to Chrome's Task Manager, the tab starts out with a memory footprint of about 200MB and steadily grows to ~1GB after processing ~7000 iterations and then crashes with the same OOM error:

![image](https://user-images.githubusercontent.com/1167575/159174498-9a32c63a-645a-42ea-9412-aa167da04f8f.png)

**Comment by kleisauke** [comment]
I could reproduce this and was able to fix it with commit https://github.com/kleisauke/wasm-vips/commit/89e167fcb9118d1a37f2530032fe02310dbc0eab. [Here's an updated playground link](https://wasm-vips.kleisauke.nl/playground/?disableAutoDelete=1&deflate=hVLva9swEP1XDrMRmXpy4mxf0jiwFAaj2z6UfasLdeyzfcyWjCw3dCH_-05yEpbBqMA_JN279-7dPQaFVoOFXat3kEK-z8lChbZoxKyxth9WcUySuno0stBd_HA_fNM_trJX9SyCAxis0Bg0K5jN4BhK26ASBtINGJkbk79ux4ojRBjeZipTlTYgWrRATLa45c86hU9zXvx_cxPCIVPAaxJlcKDfWG4nbTsqRn4e_OFnVd6hsmjujO6FU-8IHJQqEATvYTGfQ5qmMA99Nt2ibHUtnt8d6AjU5TUO0Btd4DBg-ezQR69wVIUlrd6ku5ZK3YIlvlA_yK8ut1S4_2J0dyr_LPBvyFDkLTIoST5CDN9z28iOlOBMskGqGxu5rHJPpW3OxZ3JEsa5y8kh4VO5dvxCo7BdTULu_Ub6Mrg3_9C3WFnOwnzJxAEfnJSQtSRXbFb357hJ1_8CqVt6WYksnEuOIHLoyIX7V3gtQY928sfDlnJvyOJPffJs5ofs0lautkQeHRShMya52i0vu1O4QTsaBdwFcOMjHi9cT84n-9ojj6wfg9jRTP4cgyhYl_QCVKZZwJB-tFmwWcd8tskU3wZPfwA) that now runs steady for me. Thanks for reporting this!

Regarding, `vips.shutdown()`, I would use it as a last resort on the web. I'm not sure if it's safe to restart libvips in WebAssembly, as [`vips_init()`](https://www.libvips.org/API/current/libvips-vips.html#vips-init) can't be called after [`vips_shutdown()`](https://www.libvips.org/API/current/libvips-vips.html#vips-shutdown). That's probably why the browser tab crashes.

As an aside, commit https://github.com/kleisauke/wasm-vips/commit/02b9215a3768d916e8e2edb94991b168a7ec76ef replaces the `EMBIND_AUTOMATIC_DELETELATER` property with the `setAutoDeleteLater` function. I updated my previous comments to reflect this.

**Comment by josephrocca** [comment]
Great, looking good so far on my end. Thanks for working to fix that so quickly!

It would definitely be nice to have some sort of `vips.destroy()`/`vips.dispose()` method that completely cleans up and removes `wasm-vips` in a "safe" way. This would be especially important in long-running single-page applications where, e.g. `wasm-vips` may only be required temporarily for a small part of the app's functionality.

(P.S. Thanks for that sneaky `scale` factor simplification ^^)

**Comment by kleisauke** [comment]
v0.0.2 now available with a fix for this. Docs for memory management have yet to be created.
